appserver/deployment/dol/src/main/java/com/sun/enterprise/deployment/util/ResourceValidator.java

```diff
@@ -0,0 +1,681 @@
+            accept((ResourceEnvReferenceDescriptor) next, env, namespace);
+        }
+
+        for (Object next : managedBean.getMessageDestinationReferenceDescriptors()) {
+            accept((MessageDestinationReferenceDescriptor) next, env, namespace);
+        }
+    }
+
+    private void accept(MessageDestinationDescriptor msgDest, JndiNameEnvironment env, JNDINamespace namespace) {
+        validateJNDIRefs(msgDest.getJndiName(), env, namespace);
+    }
+
+    /**
+     * Strategy for validating a given jndi name
+     * 1) Check in domain.xml
+     * 2) Check in the resources defined within the app. These have not been binded to the namespace yet.
+     * 3) Check for resources defined by an earlier application.
+     *
+     * In case a null jndi name is passed, we fail the deployment.
+     *
+     * @param jndiName to be validated.
+     */
+    private void validateJNDIRefs(String jndiName, JndiNameEnvironment env, JNDINamespace namespace) {
+        if (jndiName == null || "".equals(jndiName)) {
+            deplLogger.log(Level.SEVERE, RESOURCE_REF_JNDI_LOOKUP_FAILED,
+                    new Object[] {null});
+            throw new DeploymentException("Null JNDI resource");
+        }
+        if(!isResourceInDomainXML(jndiName) && !isDefaultResource(jndiName)) {
+            // convert comp to module if req
+            String convertedJndiName = getLogicalJNDIName(jndiName, env);
+            if (!namespace.find(convertedJndiName, env)) {
+                // Do a context lookup only if we are on the correct instance - Cluster case
+                try {
+                    if(loadOnCurrentInstance()) {
+                        InitialContext ctx = new InitialContext();
+                        ctx.lookup(jndiName);
+                    }
+                } catch (NamingException e) {
+                    deplLogger.log(Level.SEVERE, RESOURCE_REF_JNDI_LOOKUP_FAILED,
+                            new Object[] {jndiName});
+                    DeploymentException de = new DeploymentException(String.format("JNDI resource not present: %s", jndiName));
+                    de.initCause(e);
+                    throw de;
+                }
+            }
+        }
+    }
+
+    /**
+     * Validate the given resource in the corresponding target using domain.xml server beans.
+     * For resources defined outside the application.
+     *
+     * @param jndiName to be validated
+     * @return True if resource is present in domain.xml in the corresponding target. False otherwise.
+     */
+    private boolean isResourceInDomainXML(String jndiName) {
+        if (jndiName == null)
+            return false;
+
+        Server svr = domain.getServerNamed(target);
+        if (svr != null) {
+            return svr.isResourceRefExists(jndiName);
+        }
+
+        Cluster cluster = domain.getClusterNamed(target);
+        return cluster != null && cluster.isResourceRefExists(jndiName);
```

Comment:
Same question here for the enabled attribute. If a resource is deployed but disabled on certain target, it will not be available.